Gemfile

```diff
@@ -14,6 +14,8 @@ gem "coffee-rails", "~> 4.2"
 gem "turbolinks", "~> 5"
 gem "jbuilder", "~> 2.5"
 gem "bootstrap-sass", "3.3.7"
+gem "config"
+gem "bcrypt", "3.1.11"
```

Comment:
gem này dùng để làm gì? nói cho anh xem nào

Reply:
Ông bà mình bảo cái tốt khoe ra, xấu xa đậy lại, trong cái list đậy ấy có cả thằng password,nếu lưu password trực tiếp dưới dạng string  trong DB rất nguy hiểm gây ra lỗi bảo mật. Nếu chẳng may dữ liệu bị đánh cắp, có nhiều người dùng chỉ dùng một password cho tất cả tài khoản của họ (email, tài khoản ngân hàng) . Có cầu ắt có cung, Brypt sinh ra để giải quyết vấn đề này. brypt() là hasing algorithm (cái gem brypt viết dựa trên thuật toán ấy -_- ). Nó nhận vào một đoạn string rồi tạo vân tay số hay mã hash của đoạn string ấy. Mã hash ko thể đảo ngược .

  my_password = BCrypt::Password.create("my password")
#=> "$2a$10$vI8aWBnW3fID.ZQ4/zo1G.q1lRps.9cGLcZEiGDMVr5yUP1KUOYTa"

my_password = BCrypt::Password.new("$2a$10$vI8aWBnW3fID.ZQ4/zo1G.q1lRps.9cGLcZEiGDMVr5yUP1KUOYTa")
my_password == "my password"     #=> true
my_password == "not my password" #=> false

Nếu dùng nó cho password thì dù dữ liệu có mất cắp, thì attacker cũng ko biết được mật khẩu của user. Nhờ vậy mà Rainbow Tables ra đời, để mò password. Để giải quyết Rainbow Tables, người ta thêm một đoạn salt - data ngẫu nhiên vào password trước khi nó được  hash, và được lưu trữ  cùng hash trong DB. Nó làm tăng kích thước của Rainbow Tables, làm attacker phải sử dụng phương pháp chậm hơn là cho tất cả các mật khẩu có khả năng cho mỗi mật khẩu cá nhân
 Kết luận: Bcrypt dùng để hash dữ liệu, thường sử dụng cho password và ko bao giờ được lưu password trực tiếp vào DB :)))